docs: add blog (#2625)
apify/crawlee: website/blog/2024/08-20-problems-in-web-scraping/index.md

Thus, `Scrapy` is a good and proven solution for sites that are not protected against web scraping. You will need to develop and add the necessary solutions to the framework in order to bypass anti-scraping measures.

### [Botasaurus](https://www.omkar.cloud/botasaurus/)

A new framework for web scraping using browser automation, built on [`botasaurus-driver`](https://github.com/omkarcloud/botasaurus-driver). The initial commit was made on May 9, 2023.

Let's start with its advantages:

- Allows you to bypass any Claudflare protection as well as many others using `botasaurus-driver`.
- Good documentation for a quick start

Downsides include:

- Browser automation only, not intended for HTTP clients.
- Tight coupling with `botasaurus-driver`; you can't easily replace it with something better if it comes out in the future.
- No asynchrony, only multithreading.
- At the moment, it's quite raw and still requires fixes for stable operation.
- There are very few training materials available at the moment.

This is a good framework for quickly building a web scraper based on browser automation. It lacks flexibility and support for HTTP clients, which is crutias for users like me.

> ### [Crawlee for Python](https://www.crawlee.dev/python/)

Author: yeah, this post won't be just posted on Crawlee Blog, it will be published on Dev.to as well, also this kind of blog attracts new devs that won't know about Crawlee, they just come to read this blog and then might discover crawlee. 

so my answer is yes, it is kind of important :)

> ### [Crawlee for Python](https://www.crawlee.dev/python/)

Author: I get it. It is just an honest review of a crawlee for python developer/contributor. It does not say anything bad, though.

@mnmkng wdyt about this?

> ### [Crawlee for Python](https://www.crawlee.dev/python/)

Reviewer: I think it's alright. I agree with @barjin's points, but I think the commentary is well balanced and does not say anything bad about Crawlee for Python that wouldn't be true 😅

I believe developers will appreciate the honesty. It could also help manage expectations (underpromise, overdeliver), and also drive interest in early adopters who like new libraries where they can directly contribute to their development and help shape their future.

Btw, we add notes like this to community contributed blogs:
![CleanShot 2024-08-21 at 09 57 33@2x](https://github.com/user-attachments/assets/1a7c5938-20a3-49d9-ad37-e0b0a95d2afa)

@souravjain540 Maybe we could add a similar note, or an "editor's note", saying something like "Crawlee team: Yeah, for sure we will keep improving Crawlee for Python for years to come." Or smth like that.

---
slug: common-problems-in-web-scraping
title: 'Current problems and mistakes of web scraping in Python and tricks to solve them!'
tags: [community]

> description: 'Current problems and mistakes that developers encounters while scraping and crawling the internet with the advises and solution from an web scraping expert.'

Reviewer 2: ```suggestion
description: 'Current problems and mistakes that developers encounter while scraping and crawling the internet with advice and solutions from an web scraping expert.'
```

image: ./img/problems-in-scraping.webp
author: Max
authorTitle: Community Member of Crawlee and web scraping expert
authorURL: https://github.com/Mantisus
authorImageURL: https://avatars.githubusercontent.com/u/34358312?v=4
---

## Introduction

Greetings! I'm [Max](https://apify.com/mantisus), a Python developer from Ukraine, a developer with expertise in web scraping, data analysis, and processing.

> My journey in web scraping started in 2016 when I was solving lead generation challenges for a small company. Initially, I used off-the-shelf solutions such as [Import.io](https://www.import.io/) and Kimono Labs. However, I quickly encountered limitations such as blocking, inaccurate data extraction, and performance issues. This led me to learn Python. Those were the glory days when [`requests`](https://requests.readthedocs.io/en/latest/) and [`lxml`](https://lxml.de/)/[`beautifulsoup`](https://beautiful-soup-4.readthedocs.io/en/latest/) were enough to extract data from most websites. And if you knew how to work with threads, you were already a respected expert :)

Reviewer 2: Very minor thing, but shouldn't we replace classic emoticons with emoji?

```

It's not an error - it's a perfectly valid server response. It's encoded somehow.

The answer lies in the `Accept-Encoding` header. In the example above, I just copied it from my browser, so it lists all the compression methods my browser supports: "gzip, deflate, br, zstd". The Wayfair backend supports compression with "br", which is [Brotli](https://github.com/google/brotli), and uses it as the most efficient method.

This can happen if none of the libraries listed above have a `Brotli` dependency among their standard dependencies. However, they all support decompression from this format if you already have `Brotli` installed.

Therefore, it's sufficient to install the appropriate library:

```bash
pip install Brotli
```

This will allow you to get the result of the print:

```bash
b'<!DOCTYPE '
```

You can obtain the same result for `aiohttp` and `httpx` by doing the installation with extensions:

```bash
pip install aiohttp[speedups]
pip install httpx[brotli]
```

By the way, adding the `brotli` dependency was my first contribution to [`crawlee-python`](https://github.com/apify/crawlee-python). They use `httpx` as the base HTTP client.

> You may have also noticed that a new supported data compression format [`zstd`](https://github.com/facebook/zstd) appeared some time ago I haven't seen any backends that use it yet, but `httpx` will support decompression in versions above 0.28.0. I already use it to compress server response dumps in my projects; it shows incredible efficiency in asynchronous solutions with [`aiofiles`](https://github.com/Tinche/aiofiles).

Reviewer 3: ```suggestion
You may have also noticed that a new supported data compression format [`zstd`](https://github.com/facebook/zstd) appeared some time ago. I haven't seen any backends that use it yet, but `httpx` will support decompression in versions above 0.28.0. I already use it to compress server response dumps in my projects; it shows incredible efficiency in asynchronous solutions with [`aiofiles`](https://github.com/Tinche/aiofiles).
```

- Good documentation for a quick start

Downsides include:

- Browser automation only, not intended for HTTP clients.
- Tight coupling with `botasaurus-driver`; you can't easily replace it with something better if it comes out in the future.
- No asynchrony, only multithreading.
- At the moment, it's quite raw and still requires fixes for stable operation.
- There are very few training materials available at the moment.

This is a good framework for quickly building a web scraper based on browser automation. It lacks flexibility and support for HTTP clients, which is crutias for users like me.

### [Crawlee for Python](https://www.crawlee.dev/python/)

A new framework for web scraping in the Python ecosystem. The initial commit was made on Jan 10, 2024, with a release in the media space on July 5, 2024.

:::tip
If you like the blog so far, please consider [giving Crawlee a star on GitHub](https://github.com/apify/crawlee), it helps us to reach and help more developers.
:::

Developed by [Apify](https://apify.com/), it is a Python adaptation of their famous JS framework [`crawlee`](https://github.com/apify/crawlee), first released on Jul 9, 2019.

> As this is a completely new solution on the market, it is now in an active design and development stage. The community is also actively involved in its development. So,we can see that the use of [curl_cffi](https://github.com/apify/crawlee-python/issues/292) is already being discussed. The possibility of creating their own Rust-based client was [previously discussed](https://github.com/apify/crawlee-python/issues/80). I hope the company doesn't abandon the idea.

Reviewer 2: ```suggestion
As this is a completely new solution on the market, it is now in an active design and development stage. The community is also actively involved in its development. So, we can see that the use of [curl_cffi](https://github.com/apify/crawlee-python/issues/292) is already being discussed. The possibility of creating their own Rust-based client was [previously discussed](https://github.com/apify/crawlee-python/issues/80). I hope the company doesn't abandon the idea.
```

Also, curl_cffi support is already out.

:::note Crawlee team
"Yeah, for sure we will keep improving Crawlee for Python for years to come."
:::

As I personally would like to see an HTTP client for Python developed and maintained by a major company. And Rust shows itself very well as a library language for Python. Let's remember at least [`Ruff`](https://docs.astral.sh/ruff/) and [`Pydantic`](https://docs.pydantic.dev/latest/) v2.

Advantages:

The framework was developed by an established company in the web scraping market, which has well-developed expertise in this sphere.
- Support for both browser automation and HTTP clients.
- Fully asynchronous, based on `asyncio`.
- Active development phase and media activity. As developers listen to the community, it is quite important in this phase.

> On a separate note, it has a pretty good modular architecture. If developers introduce the ability to switch between several HTTP clients, we will get a rather flexible framework that allows us to easily change the technologies used, with a simple implementation from the development team.

Reviewer 2: It is already possible to switch HTTP clients. But it's probably fair to keep the blog this way since it's not documented and we will probably change the API soon :smile:
